csharp/src/Google.Protobuf/WritingPrimitives.cs

```diff
@@ -0,0 +1,568 @@
+        {
+            WriteRawVarint32(ref buffer, ref state, (uint)length);
+        }
+
+        #endregion
+
+        #region Writing primitives
+        /// <summary>
+        /// Writes a 32 bit value as a varint. The fast route is taken when
+        /// there's enough buffer space left to whizz through without checking
+        /// for each byte; otherwise, we resort to calling WriteRawByte each time.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteRawVarint32(ref Span<byte> buffer, ref WriterInternalState state, uint value)
+        {
+            // Optimize for the common case of a single byte value
+            if (value < 128 && state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)value;
+                return;
+            }
+
+            while (value > 127 && state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)((value & 0x7F) | 0x80);
+                value >>= 7;
+            }
+            while (value > 127)
+            {
+                WriteRawByte(ref buffer, ref state, (byte)((value & 0x7F) | 0x80));
+                value >>= 7;
+            }
+            if (state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)value;
+            }
+            else
+            {
+                WriteRawByte(ref buffer, ref state, (byte)value);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteRawVarint64(ref Span<byte> buffer, ref WriterInternalState state, ulong value)
+        {
+            while (value > 127 && state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)((value & 0x7F) | 0x80);
+                value >>= 7;
+            }
+            while (value > 127)
+            {
+                WriteRawByte(ref buffer, ref state, (byte)((value & 0x7F) | 0x80));
+                value >>= 7;
+            }
+            if (state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)value;
+            }
+            else
+            {
+                WriteRawByte(ref buffer, ref state, (byte)value);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteRawLittleEndian32(ref Span<byte> buffer, ref WriterInternalState state, uint value)
+        {
+            if (state.position + 4 > state.limit)
+            {
+                WriteRawByte(ref buffer, ref state, (byte)value);
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 8));
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 16));
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 24));
+            }
+            else
+            {
+                buffer[state.position++] = ((byte)value);
```

Comment:
This might be faster:

`BinaryPrimitives.WriteUInt32LittleEndian(buffer.Slice(state.position), value);`

If it isn't we should at least comment on why it isn't used.

Reply:
switched to WriteUInt32LittleEndian